### ISSUE TITLE
Idea: Probot for model CI/CD using Kubeflow Components

### ISSUE THREAD

**Issue by hamelsmu** [kubeflow/kubeflow#2288]
**TLDR:   Ensure model code does not cause a regression in model performance before merging with the master (or other branch).**

**Pre-Work: Wandb/Katib GitHub Probot:**

- Use an existing CI/CD tool (We are using Azure Pipelines, which are free for open source and you get one worker for private project), to test whether there is a run in [wandb](https://app.wandb.ai/) that = HEAD of the branch you are trying to merge.  If not, raise an error and fail the test. 
- If the run exists, then hit the [wandb API](https://docs.wandb.com/docs/integrations/api.html) and create a markdown table comparing the current code to the last known best run.  Let the end user define a script that has the following spec:
Inputs: wandb project, commit-id
- Output: markdown table    (I will use something like tabulate with pandas to accomplish this using the wandb python api). 
- Have the probot Dump this markdown table into the PR.
- [Optional] : have test that automatically compares previous “best” run to HEAD and is within an acceptable error bound and pass/fail the test accordingly. 
Feature: allow user to trigger this check to happen by issuing a command in the PR “/wandb check results”, and additionally poll wandb periodically to check for run that corresponds with HEAD.  

**Argo :**

In addition to this wandb probot, I want to build an Argo probot that allows me to issue a command “/argo run” on a PR to actually run the model on a GPU and push results into wandb or Katib.  This is a webhook that allows me run the model on my private infrastructure by issuing a command in a PR.  If I had this, I would integrate this with the step above for model continous integration.  

**Workflow**:
- the data scientist makes changes to the model , but needs to prove that the model performs equally or better than the incumbent model before pushing their code to master.  
- It is common in the process of modeling to have multiple experiments on improving the model architecture, some of which add value and other that do not.  While the experiments are being conducted, that code is kept on a seperate branch until the time it is ready to be "promoted" into the deployed model (however, many experiments may not be successful so not all models will be promoted.)
- even if it takes several days to train a model, ultimately you want to know the results the model is getting on the test set before deploying a new model architecture. 

cc: @jlewi

**Comment by jlewi** [comment]
Thanks @hamelsmu Can you describe your workflow in terms of PRs, branches, and model rollouts?

How long does it take to train a model and measure its efficacy? Are you worried about slowing development by doing this on every presubmit?

**Comment by hamelsmu** [comment]
@jlewi I updated the description and included a **workflow** section per your suggestion

**Comment by hamelsmu** [comment]
CI for notebooks doesn't sound like a compelling thing for me in the general case, only because of the adhoc nature of notebooks and you intentionally might want to run cells out of order etc.  

I think if you really want to run notebook as CI, you can just convert the notebook to a script using `nbconvert` and then use traditional CI (you can do this automatically in the CI system itself).  

TLDR; I am not sure you need a "special" CI for notebooks.  Were you thinking about something specifically that maybe I am not considering?

**Comment by jlewi** [comment]
@hamelsmu Good point.

This discussion got me thinking about how we could make it easy for people to find models and solutions  in GitHub

For example
https://github.com/SummitKwan/transparent_latent_gan

There's two connections to CI

1. CI is a great incentive for people to add structured metadata to their repo (e.g. travis, circleci, prow)
1. It would be good to know whether solutions work or not.

I wonder if we could come up with a useful way to decoratively describe how to test a model such that we could extract useful information about what a model is doing.

**Comment by hamelsmu** [comment]
That’s interesting.  Would you mind expanding your thoughts on what you
have in mind with decorative aspect of it?  What do you mean by
decoratively describe?

**Comment by jlewi** [comment]
I don't have concrete ideas; just some notion that it would be some sort of YAML file that would ultimately do 2 things

1. Make it easy to spin up the solution on Kubeflow
1. Provide some meta information e.g. description, keywords etc... to make it easy to search GitHub.

Lets take you're GitHub Issue summarization example in kubeflow. We basically have a set of templatized YAML specs in the form of ksonnet app that make it easy to deploy on Kubeflow
https://github.com/kubeflow/examples/tree/master/github_issue_summarization/ks_app

But that's not easily discoverable/machine readable.

But we do have a CI file at the root of the repo which is pure YAML
https://github.com/kubeflow/examples/blob/master/prow_config.yaml

So this is very machine readable. But there's no meta information to indicate that anything there would be useful as an ML/Kubeflow example.

In Kubernetes there is an Application spec intended to describe applications and it can contain keywords etc....

We plan on leveraging those in Kubeflow. So we could add an Application spec to each of our examples to provide some metainformation.

If we then establish some convention akin to a .gitignore file or a travis.yaml file its very easy to mine public repos to identify solutions.

But we need some incentive for people to provide that information. CI seems like a huge incentive.
